Translate concepts/overview/working-with-objects/labels.md in Korean (#15831)
kubernetes/website: content/ko/docs/concepts/overview/working-with-objects/labels.md

---
title: 레이블과 셀렉터
content_template: templates/concept
weight: 40
---

{{% capture overview %}}

> _레이블_은 파드와 같은 오브젝트에 첨부된 키와 값의 쌍이다.

Reviewer: Please add a space like `_레이블_ 은` because of wrong rendering on https://deploy-preview-15831--k8s-dev-ko.netlify.com/ko/docs/concepts/overview/working-with-objects/labels/ like the screenshot.
<img width="404" alt="image" src="https://user-images.githubusercontent.com/6965982/63320010-36c59400-c357-11e9-93f0-f151aa46d07a.png">

Reviewer 2: 아차.. 띄어쓰기 해주시며 되겠네요.
```suggestion
_레이블_ 은 파드와 같은 오브젝트에 첨부된 키와 값의 쌍이다.
```

레이블은 오브젝트의 특성을 식별하는 데 사용되어 사용자에게 중요하지만, 코어 시스템에 직접적인 의미는 없다.
레이블로 오브젝트의 하위 집합을 선택하고, 구성하는데 사용할 수 있다. 레이블은 오브젝트를 생성할 때에 붙이거나 생성 이후에 붙이거나 언제든지 수정이 가능하다.
오브젝트마다 키와 값으로 레이블을 정의할 수 있다. 오브젝트의 키는 고유한 값이어야 한다.

```json
"metadata": {
  "labels": {
    "key1" : "value1",
    "key2" : "value2"
  }
}
```

레이블은 UI와 CLI에서 효율적인 쿼리를 사용하고 검색에 사용하기에 적합하다. 식별되지 않는 정보는 [어노테이션](/docs/concepts/overview/working-with-objects/annotations/)으로 기록해야 한다.

{{% /capture %}}

{{% capture body %}}

> ## 동기

Reviewer: This is just a comment and I would like to listen to the comment from the author on using '동기'. In Korean, just '동기' can be mis-understood as 'synchronous'. How about translating liberally like '사용 동기'?

Author: 제안해주신 '사용 동기'가 '동기'보다는 더 나은 표현인 것 같습니다.
해당 제안 반영하겠습니다 :)

레이블 예시는 일반적으로 사용하는 경우에 해당한다. 당신의 규약에 따라 자유롭게 개발할 수 있다. 오브젝트에 붙여진 레이블 키는 고유해야한다는 것을 기억해야한다.

## 구문과 캐릭터 셋

_레이블_ 은 키와 값의 쌍이다. 유효한 레이블 키에는 슬래시(`/`)로 구분되는 선택한 접두사와 이름이라는 2개의 세그먼트가 있다. 이름 세그먼트는 63자 미만으로 시작과 끝은 알파벳과 숫자(`[a-z0-9A-Z]`)이며, 대시(`-`), 밑줄(`_`), 점(`.`)과 함께 사용할 수 있다. 접두사는 선택이다. 만약 접두사를 지정한 경우 접두사는 DNS의 하위 도메인으로 해야하며, 점(`.`)과, 전체 253자 이하, 슬래시(`/`)로 구분되는 DNS 레이블이다.

접두사를 생략하면 키 레이블은 개인용으로 간주한다. 최종 사용자의 오브젝트에 자동화된 시스템 구성 요소(예: `kube-scheduler`, `kube-controller-manager`, `kube-apiserver`, `kubectl` 또는 다른 타사의 자동화 구성 요소)의 접두사를 지정해야 한다.

`kubernetes.io/`와 `k8s.io/` 접두사는 쿠버네티스의 핵심 구성요소로 예약되어있다.

유효한 레이블 값은 63자 미만 또는 공백이며 시작과 끝은 알파벳과 숫자(`[a-z0-9A-Z]`)이며, 대시(`-`), 밑줄(`_`), 점(`.`)과 함께 사용할 수 있다. 

다음의 예시는 파드에 `environment: production` 과 `app: nginx` 2개의 레이블이 있는 구성 파일이다.

```yaml

apiVersion: v1
kind: Pod
metadata:
  name: label-demo
  labels:
    environment: production
    app: nginx
spec:
  containers:
  - name: nginx
    image: nginx:1.7.9
    ports:
    - containerPort: 80
    
```

## 레이블 셀렉터

[이름과 UID](/docs/user-guide/identifiers)와 다르게 레이블은 고유하지 않다. 일반적으로 우리는 많은 오브젝트에 같은 레이블을 가질 것으로 예상한다.

레이블 셀렉터를 통해 클라이언트와 사용자는 오브젝트를 식별할 수 있다. 레이블 셀렉터는 쿠버네티스 코어 그룹의 기본이다.

> API는 현재 _일치성 기준_과 _집합성 기준_ 이라는 두 종류의 셀렉터를 지원한다.

Reviewer: Please purposely add a space like `_일치성 기준_ 과` because of wrong rendering like the attached screenshot:
<img width="337" alt="image" src="https://user-images.githubusercontent.com/6965982/63320225-09c5b100-c358-11e9-9615-3b5142c893db.png">

Reviewer 2: ```suggestion
API는 현재 _일치성 기준_ 과 _집합성 기준_ 이라는 두 종류의 셀렉터를 지원한다.
```

---
reviewers:

> - mikedanese

Author: 원문 리뷰어 삭제 하겠습니다 :)

title: 레이블과 셀렉터
content_template: templates/concept
weight: 40
---

{{% capture overview %}}

> _레이블_ 은 파드와 같은 오브젝트에 붙여진 것으로 키와 값이 한 쌍으로 되어있다.

Reviewer 2: ```suggestion
_레이블_은 파드와 같은 오브젝트에 첨부된 키와 값의 쌍이다.
```
띄어쓰기와 자연스럽게 수정했습니다.

레이블 예시는 일반적으로 사용하는 경우에 해당한다. 당신의 규약에 따라 자유롭게 개발할 수 있다. 오브젝트에 붙여진 레이블 키는 고유해야한다는 것을 기억해야한다.

## 구문과 캐릭터 셋

_레이블_ 은 키와 값의 쌍이다. 유효한 레이블 키에는 슬래시(`/`)로 구분되는 선택한 접두사와 이름이라는 2개의 세그먼트가 있다. 이름 세그먼트는 63자 미만으로 시작과 끝은 알파벳과 숫자(`[a-z0-9A-Z]`)이며, 대시(`-`), 밑줄(`_`), 점(`.`)과 함께 사용할 수 있다. 접두사는 선택이다. 만약 접두사를 지정한 경우 접두사는 DNS의 하위 도메인으로 해야하며, 점(`.`)과, 전체 253자 이하, 슬래시(`/`)로 구분되는 DNS 레이블이다.

접두사를 생략하면 키 레이블은 개인용으로 간주한다. 최종 사용자의 오브젝트에 자동화된 시스템 구성 요소(예: `kube-scheduler`, `kube-controller-manager`, `kube-apiserver`, `kubectl` 또는 다른 타사의 자동화 구성 요소)의 접두사를 지정해야 한다.

`kubernetes.io/` 와 `k8s.io/` 접두사는 쿠버네티스의 핵심 구성요소로 예약되어있다.

유효한 레이블 값은 63자 미만 또는 공백이며 시작과 끝은 알파벳과 숫자(`[a-z0-9A-Z]`)이며, 대시(`-`), 밑줄(`_`), 점(`.`)과 함께 사용할 수 있다. 

다음의 예시는 파드에 `environment: production` 과 `app: nginx` 2개의 레이블이 있는 구성 파일이다.

```yaml

apiVersion: v1
kind: Pod
metadata:
  name: label-demo
  labels:
    environment: production
    app: nginx
spec:
  containers:
  - name: nginx
    image: nginx:1.7.9
    ports:
    - containerPort: 80
    
```

## 레이블 셀렉터

[이름과 UID](/docs/user-guide/identifiers)와 다르게 레이블은 고유하지 않다. 일반적으로 우리는 많은 오브젝트에 같은 레이블을 가질 것으로 예상한다.

레이블 셀렉터를 통해 클라이언트와 사용자는 오브젝트를 식별 할 수 있다. 레이블 셀렉터는 쿠버네티스 코어 그룹의 기본이다.

> API는 현재 _균등-기반_ 과 _세트-기반_ 이라는 2개의 셀렉터를 지원한다.

Author: 말씀하신 부분에 대해 동의 합니다. :)

레이블 셀렉터는 쉼표로 구분된 다양한 _요구사항_ 에따라 만들 수 있다. 다양한 요구사항이 있는 경우 쉼표 기호가 AND(`&&`) 연산자로 구분되는 역할을 하도록 해야 한다.

비어있거나 지정되지 않은 셀렉터는 상황에 따라 달라진다.
셀렉터를 사용하는 API 유형은 유효성과 의미를 문서화 해야 한다.

{{< note >}}
레플리카 셋과 같은 일부 API 유형에서 두 인스턴스의 레이블 셀렉터는 네임스페이스 내에서 겹치지 않아야 한다. 그렇지 않으면 컨트롤러는 상충되는 명령으로 보고, 얼마나 많은 복제본이 필요한지 알 수 없다.
{{< /note >}}

> ### _균등-기반_ 요건

Reviewer 2: ```suggestion
### _일치성 기준_ 요건
```
용어를 수정했습니다.

> _균등_ 또는 _불균등-기반_ 의 요구사항으로 레이블의 키와 값의 필터링을 허용한다. 일치하는 오브젝트는 추가 레이블을 가질 수 있지만 레이블의 명시된 제약 조건을 모두 만족해야 한다.

Reviewer 2: ```suggestion
_일치성 기준_ 또는 _불일치 기준_ 의 요구사항으로 레이블의 키와 값의 필터링을 허용한다. 일치하는 오브젝트는 추가 레이블을 가질 수 있지만 레이블의 명시된 제약 조건을 모두 만족해야 한다.
```
용어를 수정했습니다.

> `=`,`==`,`!=` 3개의 연산자만 인정된다. 연산자의 처음 2개는 _균등_ (그리고 단순한 동의어이다.) 나머지는 불균등을 의미한다. 예를들면, 

Reviewer 3: 리뷰 남깁니다. 띄어쓰기 확인 바랍니다.
```suggestion
`=`,`==`,`!=` 3개의 연산자만 인정된다. 연산자의 처음 2개는 _균등_ (그리고 단순한 동의어이다.) 나머지는 불균등을 의미한다. 예를 들면, 
```

> `=`,`==`,`!=` 3개의 연산자만 인정된다. 연산자의 처음 2개는 _균등_ (그리고 단순한 동의어이다.) 나머지는 불균등을 의미한다. 예를 들면, 

Reviewer 2: ```suggestion
`=`,`==`,`!=` 이 3가지 연산자만 허용한다. 처음 두 개의 연산자의 _일치성_(그리고 단순히 동의어일 뿐임), 나머지는 _불일치_를 의미한다. 예를 들면,
```
빠진 내용을 수정하고, 용어를 수정하고, 자연스럽게 수정했습니다.

```
environment = production
tier != frontend
```

전자는 `environment`를 키로 가지는 것과 `production`를 값으로 가지는 모든 리소스를 선택한다.
후자는 `tier`를 키로 가지고, 값을 `frontend`를 가지는 리소스를 제외한 모든 리소스를 선택하고, `tier`를 키로 가지며, 값을 공백으로 가지는 모든 리소스를 선택한다.
`environment=production,tier!=frontend` 처럼 쉼표를 통해 한 문장으로 `frontend`를 제외한 `production`을 필터링할 수 있다. 

> 균등-기반 레이블의 요건에 대한 한 이용 시나리오는 파드가 노드를 선택하는 기준을 지정하는 것이다.

Reviewer 3: 리뷰 남깁니다. 확인 바랍니다.
```suggestion
균등-기반 레이블의 요건에 대한 하나의 이용 시나리오는 파드가 노드를 선택하는 기준을 지정하는 것이다.
```

> 예를 들어 다음 샘플에서는 파드가 "`accelerator=nvidia-tesla-p100`" 레이블을 가진 노드를 선택한다.

Reviewer 3: 리뷰 남깁니다. 확인 바랍니다.
```suggestion
예를 들어, 아래 샘플 파드는 "`accelerator=nvidia-tesla-p100`" 레이블을 가진 노드를 선택한다.
```

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: cuda-test
spec:
  containers:
    - name: cuda-test
      image: "k8s.gcr.io/cuda-vector-add:v0.1"
      resources:
        limits:
          nvidia.com/gpu: 1
  nodeSelector:
    accelerator: nvidia-tesla-p100
```

> ### _세트-기반_ 요건

Reviewer 2: ```suggestion
### _집합성 기준_ 요건
```

> _세트-기반_ 의 레이블 요건에 따라 값 집합을 키로 필터링할 수 있다. `in`,`notin` and `exists`(키 식별자만 해당)의 3개의 연산자를 지원한다. 예를들면, 

Reviewer 2: ```suggestion
_집합성 기준_ 레이블 요건에 따라 값 집합을 키로 필터링할 수 있다. `in`,`notin` and `exists`(키 식별자만 해당)의 3개의 연산자를 지원한다. 예를 들면,
```
용어와 띄어 쓰기를 수정했습니다.

```
environment in (production, qa)
tier notin (frontend, backend)
partition
!partition
```

첫 번째 예시에서 키가 `environment`이고 값이 `production` 또는 `qa`인 모든 리소스를 선택한다.
두 번째 예시에서 키가 `tier`이고 값이 `frontend`와 `backend`를 가지는 리소스를 제외한 모든 리소스와, 키로 `tier`를 가지고 값을 공백으로 가지는 모든 리소스를 선택한다.
세 번째 예시에서 레이블의 값에 상관없이 키가 `partition`를 포함하는 모든 리소스를 선택한다.
네 번째 예시에서 레이블의 값에 상관없이 키가 `partition`를 포함하지 않는 모든 리소스를 선택한다.
마찬가지로 쉼표는 _AND_ 연산자로 작동한다. 따라서 `partition,environment notin (qa)`와 같이 사용하면 값과 상관없이 키가 `partition`인 것과 키가 `environment`이고 값이 `qa`와 다른 리소스를 필터링할 수 있다.

> _세트-기반_ 의 레이블 셀렉터는 일반적으로 `environment=production` 과 `environment in (production)`를 같은 것으로 본다. 유사하게는 `!=`과 `notin`을 같은 것으로 본다.

Reviewer 2: ```suggestion
_집합성 기준_ 레이블 셀렉터는 일반적으로 `environment=production` 과 `environment in (production)`를 같은 것으로 본다. 유사하게는 `!=`과 `notin`을 같은 것으로 본다.
```
용어를 수정했습니다.

> _세트-기반_ 의 요건은 _균등-기반_ 요건과 조합해서 사용할 수 있다. 예를 들어 `partition in (customerA, customerB),environment!=qa`

Reviewer 2: ```suggestion
_집합성 기준_ 요건은 _일치성 기준_ 요건과 조합해서 사용할 수 있다. 예를 들어 `partition in (customerA, customerB),environment!=qa`
```
용어를 수정했습니다.

## API

### LIST와 WATCH 필터링

LIST와 WATCH 작업은 쿼리 파라미터를 사용해서 반환되는 객체 집합을 필터링하기 위해 레이블 셀럭터를 지정할 수 있다. 다음의 2가지 요건 모두 허용된다(URL 쿼리 문자열을 그대로 표기함).

>   * _불균등-기반_ 요건: `?labelSelector=environment%3Dproduction,tier%3Dfrontend`

Reviewer 2: ```suggestion
  * _불일치 기준_ 요건: `?labelSelector=environment%3Dproduction,tier%3Dfrontend`
```
용어를 수정했습니다.

>   * _세트-기반_ 요건: `?labelSelector=environment+in+%28production%2Cqa%29%2Ctier+in+%28frontend%29`

Reviewer 2: ```suggestion
  * _집합성 기준_ 요건: `?labelSelector=environment+in+%28production%2Cqa%29%2Ctier+in+%28frontend%29`
```
용어를 수정했습니다.

> 두가지 레이블 셀렉터 스타일은 모두 REST 클라이언트를 통해 리소스를 확인하거나 목록을 볼 수 있다. 예를들어 `kubectl`로 `apiserver`에 _불균등-기반_으로 하는 명령어를 다음과 같이 사용할 수 있다.

Reviewer 3: 리뷰 남깁니다. 확인 바랍니다.
```suggestion
두가지 레이블 셀렉터 스타일은 모두 REST 클라이언트를 통해 리소스를 확인하거나 목록을 볼 수 있다. 예를 들어, `kubectl`로 `apiserver`에 _불균등-기반_으로 하는 명령어를 다음과 같이 사용할 수 있다.
```

```shell
kubectl get pods -l environment=production,tier=frontend
```

또는 _세트-기반_ 요건을 사용하면

```shell
kubectl get pods -l 'environment in (production),tier in (frontend)'
```

앞서 안내한 것처럼 _세트-기반_ 요건은 더 보여준다. 예시에서 다음과 같이 OR 연산자를 구현할 수 있다.

```shell
kubectl get pods -l 'environment in (production, qa)'
```

> 또는 _존재_ 연산자로 불일치한 것을 찾아 제한할 수 있다. 

Reviewer 2: ```suggestion
또는 _exists_ 연산자에 불일치한 것으로 제한할 수 있다.
```
`존재`번역을 빼고, 자연스럽게 수정했습니다.

```shell
kubectl get pods -l 'environment,environment notin (frontend)'
```

### API 객체에서 참조 설정

> [`서비스`](/docs/user-guide/services) 와 [`리플리케이션 컨트롤러`](/docs/user-guide/replication-controller)와 같은 일부 쿠버네티스 오브젝트는 레이블 셀렉터를 사용해서 [`파드`](/docs/user-guide/pods)와 같은 다른 리소스 집합을 선택한다.

Reviewer 3: 리뷰 남깁니다. 확인 바랍니다.
```suggestion
[`서비스`](/docs/user-guide/services) 와 [`레플리케이션 컨트롤러`](/docs/user-guide/replication-controller)와 같은 일부 쿠버네티스 오브젝트는 레이블 셀렉터를 사용해서 [`파드`](/docs/user-guide/pods)와 같은 다른 리소스 집합을 선택한다.
```

서비스와 리플리케이션 컨트롤러의 레이블 셀렉터는 `json` 또는 `yaml` 파일에 매핑된 _균등-기반_ 요구사항의 셀렉터만 지원한다.

```json
"selector": {
    "component" : "redis",
}
```
or

```yaml
selector:
    component: redis
```

`json` 또는 `yaml` 서식에서 셀렉터는 `component=redis` 또는 `component in (redis)` 모두 같은 것이다.

#### 세트-기반 요건을 지원하는 리소스

[`잡`](/docs/concepts/jobs/run-to-completion-finite-workloads/), [`디플로이먼트`](/docs/concepts/workloads/controllers/deployment/), [`레플리카 셋`](/docs/concepts/workloads/controllers/replicaset/) 그리고 [`데몬 셋`](/docs/concepts/workloads/controllers/daemonset/) 같은 새로운 리소스들은 세트-기반의 요건도 지원한다.

```yaml
selector:
  matchLabels:
    component: redis
  matchExpressions:
    - {key: tier, operator: In, values: [cache]}
    - {key: environment, operator: NotIn, values: [dev]}
```

`matchLables`는 `{key,value}`의 쌍과 매칭된다. `matchLables`에 매칭된 단일 `{key,value}`는 `matchExpressions`의 요소와 같으며 `key` 필드는 "key"로, `operator`는 "In" 그리고 `values`에는 "value"만 나열되어 있다. `matchExpressions`는 파드 셀렉터의 요건 목록이다. 유효한 연산자에는 In, NotIn, Exists 및 DoNotExist가 포함된다. In 및 NotIn은 설정된 값이 있어야 한다. `matchLabels` 과 `matchExpressions` 모두 AND로 되어있어 일치하기 위해서는 모든 요건을 만족해야한다.

#### 노드 셋 선택

레이블로 선택하는 한 가지 방법은 파드가 스케줄할 수 있는 노드 셋을 제한하는 것이다.

> 자세한 내용은 [노드 선택](/docs/concepts/configuration/assign-pod-node/) 문서를 참조하라.

Reviewer 3: 리뷰 남깁니다. 확인 바랍니다.

```suggestion
자세한 내용은 [노드 선택](/docs/concepts/configuration/assign-pod-node/) 문서를 참조한다.
```
